libcalico-go/lib/backend/k8s/conversion/workload_endpoint_default.go

```diff
@@ -296,3 +329,139 @@ func HandleSourceIPSpoofingAnnotation(annot map[string]string) ([]string, error)
+		egressBurst, err := parseAndValidateQty(str, *resource.NewQuantity(qosControls.EgressBandwidth, resource.DecimalSI), maxBurst)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing egress burst annotation: %w", err))
+		} else {
+			qosControls.EgressBurst = egressBurst
+		}
+	}
+
+	// calico packet rate annotations
+	if str, found := annotations[qosIngressPacketRateAnnotation]; found {
+		ingressPacketRate, err := parseAndValidateQty(str, minPacketRate, maxPacketRate)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing ingress packet rate annotation: %w", err))
+		} else {
+			qosControls.IngressPacketRate = ingressPacketRate
+		}
+	}
+	if str, found := annotations[qosEgressPacketRateAnnotation]; found {
+		egressPacketRate, err := parseAndValidateQty(str, minPacketRate, maxPacketRate)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing egress packet rate annotation: %w", err))
+		} else {
+			qosControls.EgressPacketRate = egressPacketRate
+		}
+	}
+
+	// calico number of connections annotations
+	if str, found := annotations[qosIngressMaxConnectionsAnnotation]; found {
+		ingressMaxConnections, err := parseAndValidateQty(str, minNumConnections, maxNumConnections)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing ingress max connections annotation: %w", err))
+		} else {
+			qosControls.IngressMaxConnections = ingressMaxConnections
+		}
+	}
+	if str, found := annotations[qosEgressMaxConnectionsAnnotation]; found {
+		egressMaxConnections, err := parseAndValidateQty(str, minNumConnections, maxNumConnections)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing egress max connections annotation: %w", err))
+		} else {
+			qosControls.EgressMaxConnections = egressMaxConnections
+		}
+	}
+
+	// if burst is configured, bandwidth must be configured
+	if qosControls.IngressBurst != 0 && qosControls.IngressBandwidth == 0 {
+		errs = append(errs, fmt.Errorf("ingress bandwidth must be specified when ingress burst is specified"))
+		qosControls.IngressBurst = 0
+	}
+	if qosControls.EgressBurst != 0 && qosControls.EgressBandwidth == 0 {
+		errs = append(errs, fmt.Errorf("egress bandwidth must be specified when egress burst is specified"))
+		qosControls.EgressBurst = 0
+	}
+
+	// default burst values if bandwidth is configured
+	if qosControls.IngressBandwidth != 0 && qosControls.IngressBurst == 0 {
+		qosControls.IngressBurst = maxBurst.Value()
+	}
+	if qosControls.EgressBandwidth != 0 && qosControls.EgressBurst == 0 {
+		qosControls.EgressBurst = maxBurst.Value()
+	}
+
+	// return nil if no control is configured
+	if (*qosControls == libapiv3.QoSControls{}) {
+		qosControls = nil
+	}
+
+	return qosControls, errors.Join(errs...)
+}
+
+func parseAndValidateQty(str string, minQty, maxQty resource.Quantity) (int64, error) {
+	qty, err := resource.ParseQuantity(str)
+	if err != nil {
+		return 0, err
+	}
+
+	if qty.Value() < minQty.Value() {
```

Comment:
Might be better to clamp to min/max values (given that user will have limited feedback from the error)?